cmd/tools/genfaultinjection/main.go

```diff
@@ -0,0 +1,235 @@
+	w, err := os.Create(filename)
+	fatalIfErr(err)
+	defer func() {
+		fatalIfErr(w.Close())
+	}()
+	_, err = fmt.Fprintf(w, "%s\n// Code generated by cmd/tools/genfaultinjection. DO NOT EDIT.\n", licenseText)
+	fatalIfErr(err)
+	generator(w, store)
+}
+
+func generateFaultInjectionStore(w io.Writer, store reflect.Type) {
+	writeStoreHeader(w, store, `
+package faultinjection
+
+import (
+	"context"
+{{range .ExtraImports}}
+	{{printf "%q" .}}
+{{- end}}
+	"go.temporal.io/server/common/persistence"
+)
+
+type (
+	faultInjection{{.StoreName}} struct {
+		baseStore persistence.{{.StoreName}}
+		generator faultGenerator
+	}
+)
+
+func newFaultInjection{{.StoreName}}(
+	baseStore persistence.{{.StoreName}},
+	generator faultGenerator,
+) *faultInjection{{.StoreName}} {
+	return &faultInjection{{.StoreName}}{
+		baseStore: baseStore,
+		generator: generator,
+	}
+}
+`)
+
+	writeStoreMethods(w, store, `
+func (c *faultInjection{{.StoreName}}) {{.MethodName}}(
+{{.InParams}}) {{.OutParams}} {
+	return inject{{.NumOutParams}}(c.generator.generate("{{.MethodName}}"), func() {{.OutParams}} {
+		return c.baseStore.{{.MethodName}}({{.InVars}})
+	})
+}
+`, `
+func (c *faultInjection{{.StoreName}}) {{.MethodName}}(
+{{.InParams}}) {{.OutParams}} {
+	{{if ne .OutParams ""}}return {{end}}c.baseStore.{{.MethodName}}({{.InVars}})
+}
+`)
+}
+
+func writeStoreHeader(w io.Writer, store reflect.Type, tmpl string) {
+	fatalIfErr(template.Must(template.New("code").Parse(tmpl)).Execute(w, map[string]any{
+		"StoreName":    store.Name(),
+		"ExtraImports": extraImports[store.Name()],
+	}))
+}
+
+func writeStoreMethods(w io.Writer, store reflect.Type, tmpl string, passThroughTmpl string) {
+	for i := 0; i < store.NumMethod(); i++ {
+		writeStoreMethod(w, store, store.Method(i), tmpl, passThroughTmpl)
+	}
+}
+
+func writeStoreMethod(w io.Writer, store reflect.Type, m reflect.Method, tmpl string, passThroughTmpl string) {
+
+	mT := m.Type
+	inParams := ""
+	inVars := ""
+	for i := 0; i < mT.NumIn(); i++ {
+		pName := fmt.Sprintf("p%d", i)
+		// Special names for ctx and request for better intellisense support.
+		if mT.In(i).String() == "context.Context" {
+			pName = "ctx"
+		} else if strings.HasSuffix(mT.In(i).String(), "Request") {
+			pName = "request"
+		}
+		inParams += fmt.Sprintf("\t %s %s, \n", pName, mT.In(i).String())
+		inVars += fmt.Sprintf("%s, ", pName)
+	}
+	if len(inParams) > 2 {
+		// Remove trailing ", " if any.
+		inVars = inVars[:len(inVars)-2]
+	}
+
+	outParams := ""
+	for i := 0; i < mT.NumOut(); i++ {
+		outParams += fmt.Sprintf("%s, ", mT.Out(i).String())
+	}
+	if len(outParams) > 2 {
+		// Remove trailing ", " if any.
+		outParams = outParams[:len(outParams)-2]
+		outParams = fmt.Sprintf("(%s)", outParams)
+	}
+
+	fields := map[string]string{
+		"StoreName":    store.Name(),
+		"MethodName":   m.Name,
+		"InParams":     inParams,
+		"InVars":       inVars,
+		"OutParams":    outParams,
+		"NumOutParams": fmt.Sprintf("%d", mT.NumOut()-1),
```

Comment:
I forgot this one: you can just use an integer here (with type any). everything is automatically stringified